nav2_util/include/nav2_util/simple_action_server.hpp

```diff
@@ -77,14 +82,24 @@ class SimpleActionServer
 
   rclcpp_action::GoalResponse handle_goal(
     const rclcpp_action::GoalUUID & /*uuid*/,
-    std::shared_ptr<const typename ActionT::Goal>/*goal*/)
+    std::shared_ptr<const typename ActionT::Goal> goal)
   {
     std::lock_guard<std::recursive_mutex> lock(update_mutex_);
 
     if (!server_active_) {
       return rclcpp_action::GoalResponse::REJECT;
     }
 
+    // Only ckeck the goal if a callback was bind
+    if (goal_checker_callback_ != nullptr) {
```

Comment:
https://github.com/ros-planning/navigation2/blob/main/nav2_bt_navigator/src/bt_navigator.cpp#L283 our servers have the ability already to check if a new goal is been requested to handle. This seems like an analog that breaks the "simple" of the simple action server. One of the goals of the simple action server is to remove all of these callback functions and let us talk to an action server object.

This re-exposes one of the action server callbacks, at that point, I'd think it makes sense to just use the normal action server in this situation. I recognize that the existing implementation relies on polling for updates, so there would have to be a good & specific reason to also expose this (or remove the existing polling feature)

Reply:
But with the catch that the simple server can stay simple..?
This callback function is not a hard requirement.

But yes. It re-exposes this one feature. I find it valuable to reexpose this and place "checkers" like that in the designed position.

This would keep your robot to stop for a goal that does not meet very simple requirements. Because when you accept it you have to deal with preemtion and try to stop it at a far later point. Everything is doable, but I just like to have such checkers at the most valuable position.

Stuff about preemtion should probably live in a extra preemtion_checker space. But I believe there exists a valid space for such `action_goal_checkers`

Reply:
I think we're covering all the comment topics in the main-level discussion, let's chat there about these "big" questions and then readdress the specific comments once we have some agreement in general design